src/System.IO.Pipelines/Pipe.cs

```diff
@@ -64,36 +65,38 @@ public class Pipe : IPipelineReader, IPipelineWriter, IReadableBufferAwaiter, IF
         /// Initializes the <see cref="Pipe"/> with the specifed <see cref="IBufferPool"/>.
         /// </summary>
         /// <param name="pool"></param>
-        /// <param name="scheduler"></param>
-        /// <param name="maximumSizeLow"></param>
-        /// <param name="maximumSizeHigh"></param>
-        public Pipe(IBufferPool pool, IScheduler scheduler = null, long maximumSizeLow = 0, long maximumSizeHigh = 0)
+        /// <param name="options"></param>
+        public Pipe(IBufferPool pool, PipeOptions options = null)
         {
             if (pool == null)
             {
                 throw new ArgumentNullException(nameof(pool));
             }
 
-            if (maximumSizeLow < 0)
+            options = options ?? new PipeOptions();
+
+            if (options.MaximumSizeLow < 0)
             {
-                throw new ArgumentOutOfRangeException(nameof(maximumSizeLow));
+                throw new ArgumentOutOfRangeException(nameof(options.MaximumSizeLow));
             }
 
-            if (maximumSizeHigh < 0)
+            if (options.MaximumSizeHigh < 0)
             {
-                throw new ArgumentOutOfRangeException(nameof(maximumSizeHigh));
+                throw new ArgumentOutOfRangeException(nameof(options.MaximumSizeHigh));
             }
 
-            if (maximumSizeLow > maximumSizeHigh)
+            if (options.MaximumSizeLow > options.MaximumSizeHigh)
             {
-                throw new ArgumentException(nameof(maximumSizeHigh) + " should be greater or equal to " + nameof(maximumSizeLow), nameof(maximumSizeHigh));
+                throw new ArgumentException(nameof(options.MaximumSizeHigh) + " should be greater or equal to " + nameof(options.MaximumSizeLow), nameof(options.MaximumSizeHigh));
             }
+
             _pool = pool;
-            _scheduler = scheduler ?? InlineScheduler.Default;
-            _maximumSizeHigh = maximumSizeHigh;
-            _maximumSizeLow = maximumSizeLow;
-            _awaitableState = _awaitableIsNotCompleted;
-            _flushAwaitableState = _awaitableIsCompleted;
+            _readerScheduler = options.ReaderScheduler ?? InlineScheduler.Default;
```

Comment:
See the discussion above. Making it immutable means I need to have overloads for all combinations of fields.